Group contact self-service features

skizzerz opened the pull request:
- GCs will now be shown all group contacts, both public and unlisted, for their own projects in ChanServ INFO output.
- Online GCs will be notified by ProjectServ whenever a channel is registered inside of their namespace.
- A new command ChanServ LISTGROUPCHANS for GCs to list all channels (optionally filtered by a pattern) for their projects.
- A new command NickServ LISTGROUPCLOAKS for GCs to list all cloaks (optionally filtered by a pattern) for their projects.

skizzerz committed:
Group contact self-service features

- GCs will now be shown all group contacts, both public and unlisted,
  for their own projects in ChanServ INFO output.
- Online GCs will be notified by ProjectServ whenever a channel is
  registered inside of their namespace.
- A new command ChanServ LISTGROUPCHANS for GCs to list all channels
  (optionally filtered by a pattern) for their projects.
- A new command NickServ LISTGROUPCLOAKS for GCs to list all cloaks
  (optionally filtered by a pattern) for their projects.

aaronmdjones commented on projectns/cs_listgroupchans.c:
```diff
@@ -0,0 +1,99 @@
+
+	mowgli_node_t *n;
+	mowgli_list_t *plist = projectsvs->myuser_get_projects(si->smu);
+	unsigned int num_projects = 0;
+
+	MOWGLI_ITER_FOREACH(n, plist->head)
+		num_projects++;
+
+	if (!num_projects)
```
This can be simplified to

```C
    if (! MOWGLI_LIST_LENGTH(plist))
```

skizzerz replied:
Done, albeit without the space between the ! operator and the macro. If this is required for atheme coding conventions, I'll add that space in.

aaronmdjones commented on projectns/cs_listgroupchans.c:
```diff
@@ -0,0 +1,99 @@
+			command_success_nodata(si, ngettext(N_("\2%d\2 match for pattern \2%s\2"), N_("\2%d\2 matches for pattern \2%s\2"), matches),
+			                       matches, filter);
+		else
+			command_success_nodata(si, ngettext(N_("\2%d\2 match"), N_("\2%d\2 matches"), matches), matches);
```
`matches` is an `unsigned int`, so you should use `%u` not `%d` in both paths here.

aaronmdjones commented on projectns/cs_listgroupchans.c:
```diff
@@ -0,0 +1,99 @@
+DECLARE_MODULE_V1
+(
+		"freenode/projectns/cs_listgroupchans", MODULE_UNLOAD_CAPABILITY_OK, mod_init, mod_deinit,
+		"", "freenode <http://www.freenode.net>"
```
freenode did not author this module.

skizzerz replied:
Changed author, but left path the same for consistency with all of the other modules (and since the Makefile dumps them all into a freenode/ directory). We can clean that up, but I feel that's probably out of scope for this particular PR.

aaronmdjones commented on projectns/ns_listgroupcloaks.c:
```diff
@@ -0,0 +1,134 @@
+				ngettext(N_("\2%d\2 match for pattern \2%s\2"), N_("\2%d\2 matches for pattern \2%s\2"), matches),
+				matches, filter);
+		else
+			command_success_nodata(si, ngettext(N_("\2%d\2 match"), N_("\2%d\2 matches"), matches), matches);
```
Same here a la `%d` vs `%u`

aaronmdjones commented on projectns/ns_listgroupcloaks.c:
```diff
@@ -0,0 +1,134 @@
+
+	MOWGLI_ITER_FOREACH_SAFE(n, tn, cloak_ns_list.head)
+	{
+		mowgli_node_delete(n, &cloak_ns_list);
```
There's no need to delete nodes from a list that only exists on the stack. It will be invalidated when the function returns anyway, so whether it has head, tail and count members set is irrelevant. You still need the `sfree` and `mowgli_node_free` though.

skizzerz committed:
Address review comments

- Use proper placeholders for unsigned int in format strings
- Change author to Libera Chat. Internal path is still kept as freenode/
  in order to maintain consistency with all of the other modules. This
  can be addressed at a later point across all of these modules.
- Remove superfluous mowgli_node_delete calls for stack-allocated lists.
